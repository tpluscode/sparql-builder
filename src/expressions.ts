export { IN } from './lib/expressions/in'
export { VALUES } from './lib/expressions/values'
