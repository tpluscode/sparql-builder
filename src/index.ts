export { sparql } from '@tpluscode/rdf-string'
export { DELETE } from './lib/DeleteBuilder.js'
export type { DeleteInsertQuery as DeleteInsert, DeleteData } from './lib/DeleteBuilder.js'
export { INSERT } from './lib/InsertBuilder.js'
export type { InsertQuery as Insert, InsertData } from './lib/InsertBuilder.js'
export { ASK } from './lib/AskBuilder.js'
export type { AskQuery as Ask } from './lib/AskBuilder.js'
export { DESCRIBE } from './lib/DescribeBuilder.js'
export type { DescribeQuery as Describe } from './lib/DescribeBuilder.js'
export { CONSTRUCT } from './lib/ConstructBuilder.js'
export type { ConstructQuery as Construct } from './lib/ConstructBuilder.js'
export { SELECT } from './lib/SelectBuilder.js'
export type { SelectQuery as Select } from './lib/SelectBuilder.js'
export { WITH } from './lib/WithBuilder.js'
export { prefixes } from '@tpluscode/rdf-string'
export type { SparqlTemplateResult } from '@tpluscode/rdf-string'
export type {
  SparqlAskExecutable, SparqlGraphQueryExecutable, SparqlQueryExecutable, SparqlUpdateExecutable,
} from './lib/index.js'
