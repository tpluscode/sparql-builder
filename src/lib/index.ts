import { AskQuery, ConstructQuery, QueryOptions, SelectQuery, UpdateQuery } from 'sparql-http-client'
import { SparqlTemplateResult } from '@tpluscode/rdf-string'
import prologue, { PrologueBuilder } from './partials/prologue'

interface SparqlBuildOptions {
  base?: string
}

export type SparqlExecuteOptions = QueryOptions & SparqlBuildOptions

export interface SparqlQuery extends PrologueBuilder {
  build(options?: SparqlBuildOptions): string
  _getTemplateResult(): SparqlTemplateResult
}

export interface SparqlQueryExecutable {
  execute<TQuery extends SelectQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['select']>
}

export interface SparqlGraphQueryExecutable {
  execute<TQuery extends ConstructQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['construct']>
}

export interface SparqlUpdateExecutable {
  execute<TQuery extends UpdateQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['update']>
}

export interface SparqlAskExecutable {
  execute<TQuery extends AskQuery<any>>(client: TQuery, requestInit?: SparqlExecuteOptions): ReturnType<TQuery['ask']>
}

type TBuilder = Pick<SparqlQuery, 'build'> & Pick<SparqlTemplateResult, '_toPartialString'>

// eslint-disable-next-line no-unused-vars
export default function Builder<T extends SparqlQuery>(): TBuilder & T {
  return {
    ...prologue(),
    build(this: SparqlQuery, { base }: SparqlBuildOptions = {}): string {
      const queryResult = this._getTemplateResult().toString({
        base,
      })

      if (this.prologueResult) {
        return `${this.prologueResult}\n\n${queryResult}`
      }

      return queryResult
    },
    _toPartialString(this: SparqlQuery, options: any) {
      return this._getTemplateResult()._toPartialString(options)
    },
  } as any
}
