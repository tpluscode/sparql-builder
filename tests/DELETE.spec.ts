import { owl, schema } from '@tpluscode/rdf-ns-builders'
import { sparqlClient } from './_mocks'
import { DELETE } from '../src'

describe('DELETE', () => {
  it('adds an empty WHERE if no pattern provided', () => {
    // given
    const expected = `DELETE {
      ?s ?p ?o .
    } WHERE {}`

    // when
    const query = DELETE`?s ?p ?o .`.build()

    // then
    expect(query).toMatchQuery(expected)
  })

  it('combines multiple DELETE calls', () => {
    // given
    const expected = `PREFIX owl: <http://www.w3.org/2002/07/owl#>

    DELETE {
      <http://example.com/bar> owl:sameAs <http://example.org/bar> .
      <http://example.com/foo> owl:sameAs <http://example.org/foo> .
    } WHERE {}`

    // when
    const query = DELETE`<http://example.com/bar> ${owl.sameAs} <http://example.org/bar> .`
      .DELETE`<http://example.com/foo> ${owl.sameAs} <http://example.org/foo> .`
      .build()

    // then
    expect(query).toMatchQuery(expected)
  })

  it('has a WHERE method', () => {
    // given
    const expected = `PREFIX schema: <http://schema.org/> 
    
    DELETE {
      ?s ?p ?o .
    } WHERE {
      ?s a schema:Person ; ?p ?o
    }`

    // when
    const query = DELETE`?s ?p ?o .`
      .WHERE`?s a ${schema.Person} ; ?p ?o`
      .build()

    // then
    expect(query).toMatchQuery(expected)
  })

  it('executes as update', async () => {
    // given
    const client = sparqlClient()

    // when
    await DELETE``.execute(client)

    // then
    expect(client.update).toHaveBeenCalled()
  })
})
