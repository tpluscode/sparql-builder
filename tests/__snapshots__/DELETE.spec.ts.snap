// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE can have additional prologue 1`] = `
"#pragma join.hash off
BASE <http://foo.bar/baz>

PREFIX owl: <http://www.w3.org/2002/07/owl#>

DELETE { <http://example.com/bar> owl:sameAs <http://example.org/bar> . } WHERE {}"
`;
